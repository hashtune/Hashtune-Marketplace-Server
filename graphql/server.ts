import { ApolloServer } from 'apollo-server-express';
// import EthSigUtil from 'eth-sig-util';
import express from 'express';
import { createServer } from 'http';
import passport from 'passport';
import { createContext } from './context';
import { schema } from './schema';
var cors = require('cors');

const cookieParser = require('cookie-parser');

const { PORT = 5000 } = process.env;
const session = require('express-session');

const app = express();

// TODO move to separate file ./auth/passport.ts
// TODO use this passport middleware in mutations
app.use(session({ secret: 'secret', resave: true, saveUninitialized: true }));
app.use(cookieParser());
app.use(passport.initialize());
app.use(passport.session());
app.use(
  cors({
    origin: '*',
  })
);

app.use(express.json());
passport.serializeUser((user, done) => {
  done(null, user);
});
passport.deserializeUser((user: any, done) => {
  done(null, user);
});

const secret = process.env.SERVER_SECRET ?? '';

const JwtStrategy = require('passport-jwt').Strategy;
passport.use(
  new JwtStrategy(
    {
      jwtFromRequest: (req: any) => req.cookies['jwt'] ?? null,
      secretOrKey: secret,
    },
    async function (jwt_payload, done) {
      console.log({ jwt_payload });
      console.log('WEE');
      if (jwt_payload && jwt_payload.user) {
        return done(null, jwt_payload.user);
      } else {
        // user is false if they do not have a valid jwt
        return done(null, false);
      }
    }
  )
);

export const jwtRequired = passport.authenticate('jwt', { session: false });

app.post('/login', (req: any, res: any, next: any) => {
  passport.authenticate('jwt', async (err, user: any) => {
    console.log(req.query);
    const signedMessage = req.query.message;
    const publicKey = req.query.publicKey;

    if (user) {
      // already logged in then redirect to home
      //return next(user);
    }
    return res.json('signature and public key did not match');
  })(req, res, next);
});

app.get('/protected', jwtRequired, (req, res, next) => {
  console.log('approvied');
});

const server = createServer(app);

export const apollo = new ApolloServer({
  schema,
  introspection: true,
  apollo: {},
  context: createContext,
});

export async function main() {
  await apollo.start();
  apollo.applyMiddleware({ app });
  server.listen({ port: PORT }, () => {
    process.stdout.write(
      ` ðŸš€  Server ready at http://localhost:${PORT}${apollo.graphqlPath}`
    );
  });
}
process.env.STAGE != 'test' && main();
