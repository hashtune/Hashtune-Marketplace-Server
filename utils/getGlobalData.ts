import { prisma } from '../singletons/prisma';

export default async function getGlobalData() {
    const users = await prisma.user.findMany({});
    const artworks = await prisma.artwork.findMany({});
    if (!users || !artworks) throw new Error("Unable to set up global data");
    return { users, artworks }
}