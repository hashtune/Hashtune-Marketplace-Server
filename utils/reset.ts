import { prisma } from '../singletons/prisma';

export default async function reset() {
  try {
    await prisma.$transaction([
      prisma.bid.deleteMany({}),
      prisma.auction.deleteMany({}),
      prisma.sale.deleteMany({}),
      prisma.artwork.deleteMany({}),
      prisma.user.deleteMany({}),
    ]);
  } catch (e) {
    console.log('error tearing down tests', e);
  }
}
