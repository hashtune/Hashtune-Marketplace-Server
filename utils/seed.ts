import cryptoRandomString from 'crypto-random-string';
import { prisma } from '../singletons/prisma';

export default async function seed() {
  try {
    const generateWallet = async () => {
      const wallet = await prisma.wallet.create({
        data: {
          provider: 'metamask',
          publicKey: cryptoRandomString(20),
        },
      });
      return wallet;
    };

    const user1 = await prisma.user.create({
      data: {
        fullName: 'user 1',
        email: 'user1@dgfhjj.com',
        handle: 'user1',
        bio: "Hey I'm so and so. I've been makign artwork for x amount of years and I'm from blahblah. I love cryptooo..",
        isApprovedCreator: true,
        wallet: {
          connect: {
            id: (await generateWallet()).id,
          },
        },
      },
    });
    const user2 = await prisma.user.create({
      data: {
        fullName: 'user 2',
        email: 'user2@gmail.com',
        handle: 'user2',
        bio: "Hey I'm so an222d so. I've been makign artwork for x amount of years and I'm from blahblah. I love cryptooo..",
        isApprovedCreator: true,
        wallet: {
          connect: {
            id: (await generateWallet()).id,
          },
        },
        owned: {
          create: {
            description: 'song description',
            handle: 'art 1',
            txHash: 'abc123',
            pending: false,
            image: 'blahblah',
            link: 'blahb1lah',
            title: 'amazingsongT1itle',
            media: [{ title: 'amazingsongTitle', media: 'lala' }],
            saleType: 'auction',
            reservePrice: 50,
            creator: {
              connect: {
                id: user1.id,
              },
            },
          },
        },
      },
      include: {
        owned: true,
      },
    });
    const user3 = await prisma.user.create({
      data: {
        fullName: 'user 3',
        email: 'mr3@dgfhjj.com',
        handle: 'user3',
        bio: "Hey I'm not an approved creator and I love cryptooo..",
        isApprovedCreator: false,
        wallet: {
          connect: {
            id: (await generateWallet()).id,
          },
        },
      },
    });

    if (!user1 || !user2 || !user3) throw new Error('no Userss');

    const one = await prisma.artwork.create({
      data: {
        description: 'song description',
        handle: '1',
        txHash: 'abc123',
        pending: false,
        image: 'blahblah',
        link: 'blahblah',
        title: 'amazingsongTitle',
        media: [{ title: 'amazingsongTitle', media: 'lala' }],
        saleType: 'auction',
        reservePrice: 50,
        currentOwner: {
          connect: {
            id: user1.id,
          },
        },
        creator: {
          connect: {
            id: user1.id,
          },
        },
      },
    });
    const two = await prisma.artwork.create({
      data: {
        description: 'song description',
        handle: '2',
        txHash: 'abc123',
        pending: false,
        image: 'nothing',
        link: 'blahblah',
        title: 'amazingsongTitle',
        media: [{ title: 'amazingsongTitle', media: 'lala' }],
        saleType: 'auction',
        currentOwner: {
          connect: {
            id: user2.id,
          },
        },
        creator: {
          connect: {
            id: user2.id,
          },
        },
      },
    });
    const three = await prisma.artwork.create({
      data: {
        description: 'song description',
        handle: '3',
        txHash: 'abc123',
        pending: false,
        image: 'blahblah2',
        link: 'blahblah',
        title: 'amazingsongTitle',
        media: [{ title: 'amazingsongTitle', media: 'lala' }],
        saleType: 'auction',
        reservePrice: 50,
        auctions: {
          createMany: {
            data: [{}],
          },
        },
        currentOwner: {
          connect: {
            id: user1.id,
          },
        },
        creator: {
          connect: {
            id: user2.id,
          },
        },
      },
      include: {
        auctions: true,
      },
    });
    const four = await prisma.artwork.create({
      data: {
        description: 'song description2',
        handle: '4',
        txHash: 'abc123',
        pending: false,
        image: 'blahblah2',
        link: 'blahblah2',
        title: 'amazingsongTitle2',
        media: [{ title: 'amazingsongTitle2', media: 'lala2' }],
        saleType: 'fixed',
        price: 100,
        currentOwner: {
          connect: {
            id: user1.id,
          },
        },
        creator: {
          connect: {
            id: user1.id,
          },
        },
      },
    });
    const five = await prisma.bid.createMany({
      data: [
        {
          auctionId: three.auctions[0].id,
          offer: 25,
          userId: user1.id,
        },
      ],
    });
    const six = await prisma.artwork.create({
      data: {
        description: 'song description',
        handle: '5',
        image: 'blahblah2',
        link: 'blahblah',
        txHash: 'abc123',
        pending: false,
        title: 'amazingsongTitle',
        media: [{ title: 'amazingsongTitle', media: 'lala' }],
        saleType: 'fixed',
        price: 100,
        sales: {
          createMany: {
            data: [
              {
                price: 100,
                userId: user1.id,
              },
            ],
          },
        },
        currentOwner: {
          connect: {
            id: user1.id,
          },
        },
        creator: {
          connect: {
            id: user1.id,
          },
        },
      },
    });
    const seven = await prisma.artwork.create({
      data: {
        description: 'song description',
        handle: '6',
        image: 'blahblah2',
        link: 'blahblah',
        txHash: 'abc123',
        title: 'pending',
        media: [{ title: 'amazingsongTitle', media: 'lala' }],
        saleType: 'fixed',
        price: 100,
        sales: {
          createMany: {
            data: [
              {
                price: 100,
                userId: user1.id,
              },
            ],
          },
        },
        currentOwner: {
          connect: {
            id: user1.id,
          },
        },
        creator: {
          connect: {
            id: user1.id,
          },
        },
      },
    });
  } catch (e) {
    throw new Error(`${e}`);
  }
}
