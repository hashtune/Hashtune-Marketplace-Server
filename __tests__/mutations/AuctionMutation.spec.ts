import getGlobalData from '../../utils/getGlobalData';
import reset from '../../utils/reset';
import seed from '../../utils/seed';
import server from '../server';

describe("test auction mutation", () => {

    let users;
    let artworks;
    let auctions;

    beforeAll(async () => {
        await reset();
        await seed();
        global.testData = await getGlobalData();
        users = global.testData.users;
        artworks = global.testData.artworks;
        auctions = global.testData.auctions;
    });

    const ADD_AUCTION_MUTATION = `
        mutation addAuction2($inputType: AddAuctionInput) {
            addAuction(InputType: $inputType) {
                Auctions {
                    currentHigh
                }
                ClientErrorArtworkNotFound {
                    message
                }
                ClientErrorUserUnauthorized {
                    message
                }
                ClientErrorUnknown {
                    message
                }
                ClientErrorArtworkAlreadyExists {
                    message
                }
                ClientErrorArtworkNotAnAuction {
                    message
                }
            }
        }
    `

    const DELETE_AUCTION_MUTATION = ` 
        mutation DeleteAuction($userId: String!, $auctionId: String!) {
            deleteAuction(userId: $userId, auctionId: $auctionId) {
                Auctions {
                    currentHigh
                }
                ClientErrorAuctionNotFound {
                    message
                }
                ClientErrorUserUnauthorized {
                    message
                }
                ClientErrorUnknown {
                    message
                }
                ClientErrorArtworkNotFound {
                    message
                }
                ClientErrorAuctionNotDeletable {
                    message
                  }
            }
        }
    `

    const createAuction = async (userId, artworkId) => {
        return await server.executeOperation({
            query: ADD_AUCTION_MUTATION, variables: {
                "inputType": {
                    "userId": userId,
                    "artworkId": artworkId,
                    "live": true
                }
            }
        })
    }

    const deleteAuction = async (userId, auctionId) => {
        return await server.executeOperation({
            query: DELETE_AUCTION_MUTATION, variables: {
                "userId": userId,
                "auctionId": auctionId,
            }
        })
    }

    it("Should create a new auction", async () => {
        const res = await createAuction(users[1].id, artworks[0].id);
        expect(res).toMatchSnapshot()
    })

    it("Shouldn't let an unauthorized user create a new auction", async () => {
        const res = await createAuction(users[1].id, artworks[1].id);
        expect(res).toMatchSnapshot()
    })

    it("should fail because it cannot find the id", async () => {
        const res = await createAuction(users[1].id, "noid");
        expect(res).toMatchSnapshot()
    })

    it("Should not create a new auction for this artwork because one already exists", async () => {
        const res = await createAuction(users[0].id, artworks[3].id);
        expect(res).toMatchSnapshot()
    })

    // delete auction tests

    it("Should not find the auction", async () => {
        const res = await deleteAuction(users[1].id, "noid");
        expect(res).toMatchSnapshot()
    })

    it("Should fail to delete an auction with bids", async () => {
        const res = await deleteAuction(users[0].id, auctions[1].id);
        expect(res).toMatchSnapshot()
    })

    it("Should fail to delete an auction for an artwork that the user doesn't own", async () => {
        const res = await deleteAuction(users[2].id, auctions[0].id);
        expect(res).toMatchSnapshot()
    })

    it("Should delete an auction with no bids", async () => {
        const res = await deleteAuction(users[1].id, auctions[0].id);
        expect(res).toMatchSnapshot()
    })
})