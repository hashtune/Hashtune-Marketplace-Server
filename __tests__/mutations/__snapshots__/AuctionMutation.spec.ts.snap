// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test auction mutation Should create a new auction 1`] = `
Object {
  "data": Object {
    "addAuction": Object {
      "Auctions": Array [
        Object {
          "currentHigh": 0n,
        },
      ],
      "ClientErrorArtworkAlreadyExists": null,
      "ClientErrorArtworkNotAnAuction": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`test auction mutation Should delete an auction with no bids 1`] = `
Object {
  "data": Object {
    "deleteAuction": Object {
      "Auctions": Array [
        Object {
          "currentHigh": 0n,
        },
      ],
      "ClientErrorArtworkNotFound": null,
      "ClientErrorAuctionNotDeletable": null,
      "ClientErrorAuctionNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`test auction mutation Should fail to delete an auction for an artwork that the user doesn't own 1`] = `
Object {
  "data": Object {
    "deleteAuction": Object {
      "Auctions": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorAuctionNotDeletable": null,
      "ClientErrorAuctionNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": Object {
        "message": "The user is not the owner of the artwork",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`test auction mutation Should fail to delete an auction with bids 1`] = `
Object {
  "data": Object {
    "deleteAuction": Object {
      "Auctions": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorAuctionNotDeletable": Object {
        "message": "Cannot delete an auction with bids",
      },
      "ClientErrorAuctionNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`test auction mutation Should not create a new auction for this artwork because one already exists 1`] = `
Object {
  "data": Object {
    "addAuction": Object {
      "Auctions": null,
      "ClientErrorArtworkAlreadyExists": Object {
        "message": "An auction for the specified artwork already exists",
      },
      "ClientErrorArtworkNotAnAuction": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`test auction mutation Should not find the auction 1`] = `
Object {
  "data": Object {
    "deleteAuction": Object {
      "Auctions": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorAuctionNotDeletable": null,
      "ClientErrorAuctionNotFound": Object {
        "message": "Couldn't find an auction with the specified id",
      },
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`test auction mutation Shouldn't let an unauthorized user create a new auction 1`] = `
Object {
  "data": Object {
    "addAuction": Object {
      "Auctions": null,
      "ClientErrorArtworkAlreadyExists": null,
      "ClientErrorArtworkNotAnAuction": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": Object {
        "message": "The user is not the owner of the artwork",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`test auction mutation should fail because it cannot find the id 1`] = `
Object {
  "data": Object {
    "addAuction": Object {
      "Auctions": null,
      "ClientErrorArtworkAlreadyExists": null,
      "ClientErrorArtworkNotAnAuction": null,
      "ClientErrorArtworkNotFound": Object {
        "message": "Couldn't find an artwork with id noid",
      },
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
